### PR TITLE
test: dtpools refactor

### PR DESCRIPTION
**Pull request by gcongiu**
This PR addresses some of the points raised by Intel, such as elimination of duplicated code from DTPools tests. It also contains some restructuring of the code and cleanup for future extensions of the interface. 

As side note, I was also thinking about removing `MPI_type_dup` for basic type objects as suggested by @hajimefu. However, some tests (e.g., `test/mpi/attr/fkeyvaltype.c`) assume the framework returns derived datatypes that are eventually freed using `MPI_Type_free`. Thus for now I am not integrating this change.

### CODE REVIEW

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by wesbland** [comment]
I also have a fix I'd like to stick in here for the autotools in dtpools. I know that we warn people about building MPICH with autotools in different locations, but apparently a ton of people around here do that and have been emailing me because dtpools broke it. I believe I have a fix. Can I push it to this PR?

**Comment by wesbland** [comment]
On further consideration, it's probably orthogonal to this. I'll just open a separate PR so it can be reviewed separately.

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by gcongiu** [comment]
@wesbland if you want you can add your patch here, it should be fine.

**Comment by wesbland** [comment]
I think I'd rather leave it in a separate PR since it's orthogonal and can be merged quickly. I'd actually like to backport this into a recent internal release.

**Comment by hajimefu** [comment]
@gcongiu Thanks. I'll take a look soon.

**Comment by gcongiu** [comment]
test:jenkins/ch4/most

File: test/mpi/dtpools/src/dtpools_internal.c [hunk 1]
```diff
@@ -532,6 +532,9 @@ int DTPI_Struct_create(struct DTPI_Par *par, DTP_t dtp)
     }
 
     /* cleanup buffers */
+    if (basic_type_displs) {
```
Comment (hajimefu) [inline test/mpi/dtpools/src/dtpools_internal.c]: Actually, `free(3)` accepts `NULL` pointer (see the manpage), so you don't need this guard.

File: test/mpi/dtpools/include/dtpools_internal.h [hunk 2]
```diff
@@ -121,9 +121,94 @@ typedef enum {
     DTPI_OBJ_TYPE__SUBARRAY_C,
     DTPI_OBJ_TYPE__SUBARRAY_F,
     DTPI_OBJ_TYPE__STRUCT,
+/*
```
Comment (hajimefu) [inline test/mpi/dtpools/include/dtpools_internal.h]: The patch "test/dtpools: refactoring of object free and check functions":

It would be nice to say what was the goal of the refactoring (code movement? naming change? what else?)

File: test/mpi/include/mpitest.h [hunk 3]
```diff
@@ -58,6 +58,148 @@ typedef struct _MTestDatatype {
     int (*CheckBuf) (struct _MTestDatatype *);
 } MTestDatatype;
 
+typedef struct {
+    const char *typename;
+    MPI_Datatype type;
+} MTestType_t;
+
+#define MTEST_DEFINE_AND_INIT_TYPELIST                              \
```
Comment (hajimefu) [inline test/mpi/include/mpitest.h]: I would prefer avoiding macros if there is an alternative. This case seems to me that using extern variable for `typelist` and/or use a function for `MTEST_INIT_BASIC_SIGNATURE` would be more natural.

Comment (gcongiu) [reply test/mpi/include/mpitest.h]: Ok I will fix that

File: test/mpi/dtpools/include/dtpools_internal.h (hunk 2 above)
Comment (gcongiu) [reply test/mpi/dtpools/include/dtpools_internal.h]: I will add an explanation to the commit message. The main goal was to have for struct pools the same type of behavior of basic pools; that is, you can build derived datatypes using an arbitrary struct instead of a predefined type. I have that implemented in a separate branch but I need to do some fix up.

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/include/mpitest.h (hunk 3 above)
Comment (gcongiu) [reply test/mpi/include/mpitest.h]: @hajime I have fixed this. Unfortunately I don't have time to extend the interface now. If you want we can cherry pick just this one commit and leave the interface extension to some other time.

File: test/mpi/cxx/datatype/packsizex.cxx [hunk 4]
```diff
@@ -81,30 +40,9 @@ int main(int argc, char *argv[])
...
-                } else {
-                    cout << "Error: datatype not recognized\n";
-                    return 0;
-                }
-            } else if (!strncmp(argv[i], "-count=", strlen("-count="))) {
-                count = atoi(argv[i]+strlen("-count="));
-            }
-        }
-    }
+    err = MTestInitBasicSignatureX(argc, argv, &count, &basic_type);
```
Comment (hajimefu) [inline test/mpi/cxx/datatype/packsizex.cxx]: I assume 'X' stands for cxx and the reason for this name is to easily distinguish C++ version and C version. I like the idea, but I'd like to gather more feedback on this particular naming scheme. Is there any other use of 'X' like this?

Comment (gcongiu) [reply test/mpi/cxx/datatype/packsizex.cxx]: Yes X is for cxx. In `test/mpi/cxx` test names are suffixed by `x`. E.g., `test/mpi/cxx/attr/fkeyvaltypex.cxx`.

**Comment by gcongiu** [comment]
@hajimefu can you have another look at this? I have abandoned it for quite some time but I would like to close it.

**Comment by hajimefu** [comment]
@gcongiu Will take a look.

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

**Comment by gcongiu** [comment]
@hzhou this would mean merging the cxx and c header files, I can do that sure.

**Comment by hzhou** [comment]
@hajimefu I wonder what is the status of this PR. I am curious as PR #3401 -- adding MPIX_C_FLOAT16 datatype -- that also touches DTpools code and is waiting for this PR to get merged.

File: test/mpi/dtpools/include/dtpools_internal.h [hunk 5]
```diff
@@ -121,9 +121,94 @@ typedef enum {
...
+    DTPI_OBJ_TYPE__STRUCT_CONTIG,
+    DTPI_OBJ_TYPE__STRUCT_VECTOR,
+    DTPI_OBJ_TYPE__STRUCT_HVECTOR,
+    DTPI_OBJ_TYPE__STRUCT_BLOCK_INDEXED,
+    DTPI_OBJ_TYPE__STRUCT_BLOCK_HINDEXED,
+    DTPI_OBJ_TYPE__STRUCT_INDEXED,
+    DTPI_OBJ_TYPE__STRUCT_HINDEXED,
+    DTPI_OBJ_TYPE__STRUCT_SUBARRAY_C,
+    DTPI_OBJ_TYPE__STRUCT_SUBARRAY_F,
+*/
```
Comment (hajimefu) [inline test/mpi/dtpools/include/dtpools_internal.h]: What's the reason for having these commented names?

File: test/mpi/dtpools/src/dtpools_internal.c (hunk 1 above)
Comment (hajimefu) [reply test/mpi/dtpools/src/dtpools_internal.c]: This needs to be addressed.

File: test/mpi/dtpools/src/dtpools.c [hunk 6]
```diff
@@ -320,88 +333,16 @@ int DTP_obj_create(DTP_t dtp, int obj_idx, int val_start, int val_stride, MPI_Ai
...
-                dtp->DTP_obj_array[obj_idx].DTP_obj_buf = NULL;
-                free(dtp->DTP_obj_array[obj_idx].private_info);
-                dtp->DTP_obj_array[obj_idx].private_info = NULL;
-                break;
-            default:
-                err = DTP_ERR_OTHER;
-                fprintf(stdout, "Type structure %d is not defined\n", obj_idx);
-                fflush(stdout);
-        }
+        err = destructors[obj_type] (dtp, obj_idx);
```
Comment (hajimefu) [inline test/mpi/dtpools/src/dtpools.c]: In the old code there is a mechanism to check whether `obj_index` is within an expected range. Does the new code have a similar logic?

File: test/mpi/dtpools/include/dtpools_internal.h [hunk 7]
```diff
@@ -121,9 +121,94 @@ typedef enum {
...
+            case DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__SUBARRAY_C:            \
+                obj_type = DTPI_OBJ_TYPE__##pool_type##SUBARRAY_C;      \
+                break;                                                  \
+            case DTPI_OBJ_LAYOUT_LARGE_BLK__SUBARRAY_F:                 \
+            case DTPI_OBJ_LAYOUT_LARGE_CNT__SUBARRAY_F:                 \
+            case DTPI_OBJ_LAYOUT_LARGE_BLK_STRD__SUBARRAY_F:            \
+            case DTPI_OBJ_LAYOUT_LARGE_CNT_STRD__SUBARRAY_F:            \
+                obj_type = DTPI_OBJ_TYPE__##pool_type##SUBARRAY_F;      \
+                break;                                                  \
+            default:;                                                   \
```
Comment (hajimefu) [inline test/mpi/dtpools/include/dtpools_internal.h]: How about having a (failing) assertion here for the default case?

File: test/mpi/dtpools/src/dtpools_internal.c [hunk 8]
```diff
@@ -13,8 +13,9 @@
 #define TYPE_NAME_MAXLEN (256)
 #define BASIC_TYPE_NAME_MAXLEN (64)
 
-#define DTPI_OBJ_INIT_BUF(c_type, type_ptr)                                  \
+#define DTPI_OBJ_INIT_BUF(par, c_type, type_ptr)                             \
```
Comment (hajimefu) [inline test/mpi/dtpools/src/dtpools_internal.c]: I may be wrong, but are these changes (that move `i, j, k` into macros and add `par` as a parameter) something that should be in a different commit?

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even/debug,default

File: test/mpi/dtpools/include/dtpools_internal.h (hunk 5 above)
Comment (gcongiu) [reply test/mpi/dtpools/include/dtpools_internal.h]: They are just suggestion on possible future expansions for the struct types. I can remove them.

File: test/mpi/dtpools/include/dtpools_internal.h (hunk 7 above)
Comment (gcongiu) [reply test/mpi/dtpools/include/dtpools_internal.h]: I will fix this

File: test/mpi/dtpools/src/dtpools.c (hunk 6 above)
Comment (gcongiu) [reply test/mpi/dtpools/src/dtpools.c]: I will add the logic here too

File: test/mpi/dtpools/src/dtpools_internal.c (hunk 8 above)
Comment (gcongiu) [reply test/mpi/dtpools/src/dtpools_internal.c]: That is part of the refactoring. If you prefer I can put them in a different commit.

File: test/mpi/dtpools/include/dtpools_internal.h (hunk 5 above)
Comment (raffenet) [reply test/mpi/dtpools/include/dtpools_internal.h]: Suggestion for the future is OK as long as the comment indicates that is the case. Right now the reader does not know.

Comment (gcongiu) [reply test/mpi/dtpools/include/dtpools_internal.h]: Ok will keep that in mind for next time. Thanks

**Comment by gcongiu** [comment]
@hajimefu can you provide some information on the 48 tests that failed before I address your previous comments?

**Comment by hajimefu** [comment]
@gcongiu They are because the master branch does not have some bug fixes that we internally have. This PR should be good in terms of test results.

File: test/mpi/dtpools/src/dtpools_internal.c (hunk 8 above)
Comment (hajimefu) [reply test/mpi/dtpools/src/dtpools_internal.c]: I guess there are two solutions.
1. put this part in a different commit
2. update the commit message to add an explanation on this change

I would prefer 1., unless this part of change is strongly related with other changes (e.g. introducing function pointers for destruction/checking), to reduce the size of each patch.

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp

File: test/mpi/dtpools/src/dtpools.c (hunk 6 above)
Comment (gcongiu) [reply test/mpi/dtpools/src/dtpools.c]: Actually `obj_idx` does not need to be checked because is retrieved from the type object handler (which was previously verified as correct).

File: test/mpi/dtpools/src/dtpools_internal.c (hunk 1 above)
Comment (gcongiu) [reply test/mpi/dtpools/src/dtpools_internal.c]: Now this commit adds the logic to free the previously unfreed buffer using the same approach I was using before for other buffers. In the next commit I remove all the `if`s from every buffer that is initialized to `NULL`.

File: test/mpi/dtpools/src/dtpools_internal.c (hunk 8 above)
Comment (hajimefu) [reply test/mpi/dtpools/src/dtpools_internal.c]: > test/dtpools: minor cleanups

The commit message needs to be improved. I think it should describe a summary of changes and why it is needed.

**Comment by gcongiu** [comment]
jenkins:test/ch3/tcp
jenkins:test/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@hajimefu I rerun some tests to make sure this does not break anything. CH3 completes fine, CH4 has some fortran error that is unrelated (as far as I have understood). Let me know if there is anything else I may have missed and if not maybe we can go ahead and merge this one too.

**Comment by hajimefu** [comment]
@gcongiu Is the code based on the latest master?

**Comment by gcongiu** [comment]
@hajimefu now it is (rebased on commit 97c67ca0)

**Comment by hajimefu** [comment]
Hmm, I asked because I think I saw these errors a few weeks back, which should be already fixed in master. Let me rerun our internal test again.

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/gnu,icc/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even/debug,default

**Comment by hajimefu** [comment]
@gcongiu All failures are because the master does not have several bug fixes we've already merged internally. I think this is good to go.

**Comment by gcongiu** [comment]
@hajimefu ok, besides that I have the impression some failures may have been because my PR was not up to date with latest branch. Good to know, thanks!

**Comment by hajimefu** [comment]
@gcongiu Please sign them off on behalf of me.

**Comment by gcongiu** [comment]
@hajimefu done. Thank you for your review

**Comment by gcongiu** [comment]
test:jenkins/ch3/tcp
